02-oop-inheritance/observerPattern.js

```diff
@@ -0,0 +1,140 @@
+//Inherit function
+var inherit = function inherit(child, parent){
+  var copyOfParent = Object.create(parent.prototype);
+  copyOfParent.constructor = child;
+ 	child.prototype = copyOfParent;	
+};
+
+//Augment function
+function augment(receivingClass, givingClass, methods) {
+  if (arguments[2]) {
+    for(var i = 0, length = methods.length; i < length; i++ ) {
+      receivingClass.prototype[methods[i]] = givingClass.prototype[methods[i]];
+    }
+  } else {
+    for (var methodName in givingClass.prototype ) {
+      if (!Object.hasOwnProperty.call(receivingClass.prototype, methodName)) {
+        receivingClass.prototype[methodName] = givingClass.prototype[methodName];
+      }
+    }
+  }
+};
+
+//Observer class
+var Observer = function Observer(events){
+	this.events = events || [];
+};
+
+//Observable class
+var Observable = function Observable(){
+	this.observersList = [];
+};
+
+Observable.prototype = {
+	addObserver: function(observer){
+		this.observersList.push(observer);
+	},
+	notify: function(event,param){
+		for (var i = 0; i < this.observersList.length; i++) {
+			for (var j = 0; j < this.observersList[i].events.length; j++) {
+				if (this.observersList[i].events[j].action === event){
+					this.observersList[i].events[j].callback(this,param);
+				};
+			};
+		};
+	}
+}
+
+//Track class
+var Track = function Track (title, artist, duration){
+	Observable.call(this);
+	this.title = title;
+	this.artist = artist;
+	this.duration = duration;
+};
+
+inherit(Track,Observable);
+
+Track.prototype.get = function(attr) {
```

Comment:
Why you are defining this as multiple functions into the prototype and into the previous definition you are overwriting the entire prototype? You are not copying this code right?